### PR TITLE
DBZ 2466 Add a locking mode which doesn't conflict with DML and existing reads on Percona Server

### PR DESCRIPTION
**Pull request by insom**
https://issues.redhat.com/browse/DBZ-2466

### CODE REVIEW

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/SnapshotReader.java [hunk 1]
```diff
@@ -310,7 +310,12 @@ protected void execute() {
                 if (!snapshotLockingMode.equals(MySqlConnectorConfig.SnapshotLockingMode.NONE) && useGlobalLock) {
                     try {
                         logger.info("Step 1: flush and obtain global read lock to prevent writes to database");
-                        sql.set("FLUSH TABLES WITH READ LOCK");
+                        if (snapshotLockingMode.equals(MySqlConnectorConfig.SnapshotLockingMode.MINIMAL_PERCONA)) {
```
Comment (gunnarmorling) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/SnapshotReader.java:313]: How about providing the statement from a method on the `SnapshotLockingMode` enum?

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/SnapshotReader.java [hunk 2]
```diff
@@ -541,7 +546,8 @@ protected void execute() {
                 // ------
                 // STEP 7
                 // ------
-                if (snapshotLockingMode.equals(MySqlConnectorConfig.SnapshotLockingMode.MINIMAL) && isLocked) {
+                if ((snapshotLockingMode.equals(MySqlConnectorConfig.SnapshotLockingMode.MINIMAL) ||
```
Comment (gunnarmorling) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/SnapshotReader.java:549]: Could this be retrieved from a method on the enum, e.g. `isMinimal()`, or `requiresMinimalLocking()` or something similar?

File: debezium-connector-mysql/src/test/docker/init/setup.sql [hunk 3]
```diff
@@ -18,3 +18,20 @@ GRANT ALL PRIVILEGES ON *.* TO 'mysqluser'@'%';
 -- DATABASE:  emptydb
 -- ----------------------------------------------------------------------------------------------------------------
 CREATE DATABASE emptydb;
+RESET MASTER;
```
[the remark this answers is not in the document]
Comment (gunnarmorling) [reply debezium-connector-mysql/src/test/docker/init/setup.sql:21]: Sounds good.

File: debezium-connector-mysql/src/test/java/io/debezium/connector/mysql/SnapshotReaderIT.java [hunk 4]
```diff
@@ -224,6 +226,54 @@ private void snapshotOfSingleDatabase(boolean useGlobalLock, boolean storeOnlyMo
...
+        if (!MySQLConnection.isPerconaServer()) {
+            return; // Skip these tests for non-Percona flavours of MySQL
+        }
+
+        MySQLConnection db = MySQLConnection.forTestDatabase(DATABASE.getDatabaseName());
+        Thread t = new Thread() {
+            public void run() {
+                try {
+                    JdbcConnection connection = db.connect();
+                    connection.executeWithoutCommitting("SELECT *, SLEEP(20) FROM products_on_hand");
```
Comment (gunnarmorling) [inline debezium-connector-mysql/src/test/java/io/debezium/connector/mysql/SnapshotReaderIT.java:253]: Wow, TIL about `SLEEP` :)

File: documentation/modules/ROOT/partials/modules/mysql-connector/ref-mysql-connector-configuration-properties.adoc [hunk 5]
```diff
@@ -282,6 +282,8 @@ _Note:_ This feature should be considered an incubating one. We need a feedback
 
 `minimal` The connector holds the global read lock for just the initial portion of the snapshot while the connector reads the database schemas and other metadata. The remaining work in a snapshot involves selecting all rows from each table, and this can be done in a consistent fashion using the REPEATABLE READ transaction even when the global read lock is no longer held and while other MySQL clients are updating the database. +
 
+`minimal_percona` The connector holds the global backup lock for just the initial portion of the snapshot while the connector reads the database schemas and other metadata. The remaining work in a snapshot involves selecting all rows from each table, and this can be done in a consistent fashion using the REPEATABLE READ transaction even when the global read lock is no longer held and while other MySQL clients are updating the database. This mode does not flush tables to disk and is not blocked by long-running reads, and is only available in Percona Server +
```
Comment (gunnarmorling) [inline documentation/modules/ROOT/partials/modules/mysql-connector/ref-mysql-connector-configuration-properties.adoc:285]: Is there any description in the Percona docs where "global backup lock" could link to?

**Review by gunnarmorling — COMMENTED** [review]
Very nice; thanks, @insom! A few suggestions inline.

@jpechane, could you take a look at the CI config, should we have a new entry for the matrix build for Percona? And any thoughts on the MySql test failure? It saids there's a test failure in the logs, but number of failed tests is 0 🤔 . Just saw that the other day, too, btw.

**Comment by gunnarmorling** [comment]
Hey @insom, quick heads-up that we're going to release 1.3 Beta2 today or tomorrow. CR1 will be in a week or so, perhaps this could get wrapped up until then so we get this safely into 1.3? Thx!

**Comment by insom** [comment]
@gunnarmorling 👍 -- I got a bit derailed last week but addressing this feedback is the next thing on my todo

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/SnapshotReader.java (hunk 2 above)
Comment (insom) [reply debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/SnapshotReader.java]: I went with `usesMinimalLocking` -- the read best to me, but happy with any name, the method is a good suggestion 👍

**Comment by insom** [comment]
Found and fixed the test with the error, too -- it was because I was skipping `reader.start()` conditionally, but the `afterEach` call does an unconditional `reader.stop()` and it was causing an NPE.

**Comment by gunnarmorling** [comment]
@jpechane, can you add a commit for adjusting the CI config so it tests Percona, too? Thx!

**Comment by jpechane** [comment]
@insom Good job, applied!
@gunnarmorling There is no ned for it as when you execute `assembly` profile the Maven is configured to execute Percona tests too.

**Comment by gunnarmorling** [comment]
>  There is no need for it as when you execute assembly profile the Maven is configured to execute Percona tests too.

Interesting, I was expecting another entry in the MySQL build job matrix? How does that work?

Thanks a lot, @insom, this will be part of 1.3.0.Beta2 (due tomorrow).

**Comment by jpechane** [comment]
@gunnarmorling Failsafe plugin has multiple execution configured for assembly plugin

- MySQL
- MySQL with replica
- Percona

Those executions are run one by one

**Comment by gunnarmorling** [comment]
Aaah, gotcha. Thanks for clarifyin!

**Pull request merged**